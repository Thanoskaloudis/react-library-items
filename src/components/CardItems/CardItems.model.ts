export interface ICardItem {
  duration: string
  has_summative_assessment: boolean
  id: number
  image: string
  internal_title: string
  intro: string
  title: string
  type: string
  url: string
}


export interface ICardItems {
  items: ICardItem[]
}
