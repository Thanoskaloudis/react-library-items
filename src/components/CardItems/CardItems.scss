.cards{
  display: flex;
  flex-wrap: wrap;
  justify-content: center;

  &-item {
    flex: 0 0 25rem;
    box-sizing: border-box;
    margin: 1rem .25rem;
  }
}