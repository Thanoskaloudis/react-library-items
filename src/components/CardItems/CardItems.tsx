import React, { useEffect, useState } from 'react';
import { Card } from '../Card/Card';
import { ICardItem, ICardItems } from './CardItems.model';
import "./CardItems.scss";

export const CardItems = (props: ICardItems) => {
  const [query, setQuery] = useState("");
  const [sortState, setSortState] = useState("");
  const [showingItems, setshowingItems] = useState<ICardItem[]>([]);

  useEffect(() => {
    setshowingItems(props.items);
  }, [props])

  const updateQuery = (query: string) => {
    setQuery(query.trim());

    if(query === "") {
      setshowingItems(props.items);
    } else {
      setshowingItems(
        props.items.filter((item) =>
        item.title.toLowerCase().includes(query.toLowerCase())
      ))
    }
  };

  const updateSortState = (event: string) => {
    setSortState(event);

    if(event === "duration") {
      setshowingItems(sortItemsByDuration());
    } else if(event === "alphabetical") {
      setshowingItems(sortItemsByTitle());
    }
  };

  const sortItemsByDuration = (): ICardItem[] => {
    return showingItems.sort((a, b) => +a.duration.replace(/\D/g, "") - +b.duration.replace(/\D/g, ""));
  }

  const sortItemsByTitle = (): ICardItem[] => {
    return showingItems.sort((a, b) => {
      if (a.title > b.title) {
          return 1;
      }
      if (b.title > a.title) {
          return -1;
      }
      return 0;
    });
  }

  return (
    <div className="cards">
        <div className="cards--top">
          <input
            className="cards--top__search"
            type="text"
            placeholder="Search Library"
            value={query}
            onChange={(event) => updateQuery(event.target.value)}
          />
          <div className="cards--top__sort">
            <label className="cards--top__sort__label">Sort By</label>
            <select className="cards--top__sort__select" defaultValue={'DEFAULT'} onChange={(e) => updateSortState(e.target.value)}>
              <option value="DEFAULT" disabled>None</option>
              <option value="duration">Duration</option>
              <option value="alphabetical">Alphabetical</option>
          </select>
          </div>
      </div>
      {showingItems.map((cardItem) => (
        <div className="cards--item" key={cardItem.id}>
          <Card
            title={cardItem.title}
            imgURL={cardItem.image}
            duration={cardItem.duration}
            type={cardItem.type}
            link={cardItem.url}
            description={cardItem.intro}
          />
        </div>
      ))}
    </div>
  );
};
