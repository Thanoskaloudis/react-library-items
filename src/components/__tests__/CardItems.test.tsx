import renderer from 'react-test-renderer';
import {
  render,
  screen,
  fireEvent,
} from '@testing-library/react';
import { CardItems } from '../CardItems/CardItems';

const mockItems = [
  {
    id: 12,
    title: 'KS4',
    internal_title: 'KS4',
    url: 'https://www.blackbullion.com/pathways/key-stage-4',
    intro: '',
    duration: '9 min',
    image:
      'https://prodcontent.blackbullion.com/images/pathways/12/thumb',
    type: 'pathway',
    has_summative_assessment: false,
  },
  {
    id: 14,
    title: 'KS4',
    internal_title: 'KS4',
    url: 'https://www.blackbullion.com/pathways/key-stage-4',
    intro: '',
    duration: '9 min',
    image:
      'https://prodcontent.blackbullion.com/images/pathways/12/thumb',
    type: 'pathway',
    has_summative_assessment: false,
  },
  {
    id: 21,
    title: 'KS4',
    internal_title: 'KS4',
    url: 'https://www.blackbullion.com/pathways/key-stage-4',
    intro: '',
    duration: '9 min',
    image:
      'https://prodcontent.blackbullion.com/images/pathways/12/thumb',
    type: 'pathway',
    has_summative_assessment: false,
  },
  {
    id: 16,
    title: 'KS4',
    internal_title: 'KS4',
    url: 'https://www.blackbullion.com/pathways/key-stage-4',
    intro: '',
    duration: '9 min',
    image:
      'https://prodcontent.blackbullion.com/images/pathways/12/thumb',
    type: 'pathway',
    has_summative_assessment: false,
  },
];

describe('<Selection />', () => {
  const component = renderer.create(<CardItems items={mockItems} />);

  it('Should render correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
