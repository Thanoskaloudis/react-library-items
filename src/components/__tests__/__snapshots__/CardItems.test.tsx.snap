// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Selection /> Should render correctly 1`] = `
<div
  className="cards"
>
  <div
    className="cards--top"
  >
    <input
      className="cards--top__search"
      onChange={[Function]}
      placeholder="Search Library"
      type="text"
      value=""
    />
    <div
      className="cards--top__sort"
    >
      <label
        className="cards--top__sort__label"
      >
        Sort By
      </label>
      <select
        className="cards--top__sort__select"
        defaultValue="DEFAULT"
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="DEFAULT"
        >
          None
        </option>
        <option
          value="duration"
        >
          Duration
        </option>
        <option
          value="alphabetical"
        >
          Alphabetical
        </option>
      </select>
    </div>
  </div>
</div>
`;
